core/src/main/java/io/parsingdata/metal/data/ParseState.java

```diff
@@ -37,55 +37,61 @@
     public final BigInteger offset;
     public final Source source;
     public final ImmutableList<ImmutablePair<Token, BigInteger>> iterations;
+    public final ImmutableList<ParseReference> references;
 
-    public ParseState(final ParseGraph order, final Source source, final BigInteger offset, final ImmutableList<ImmutablePair<Token, BigInteger>> iterations) {
+    public ParseState(final ParseGraph order, final Source source, final BigInteger offset, final ImmutableList<ImmutablePair<Token, BigInteger>> iterations, final ImmutableList<ParseReference> references) {
         this.order = checkNotNull(order, "order");
         this.source = checkNotNull(source, "source");
         this.offset = checkNotNegative(offset, "offset");
         this.iterations = checkNotNull(iterations, "iterations");
+        this.references = checkNotNull(references, "references");
     }
 
     public static ParseState createFromByteStream(final ByteStream input, final BigInteger offset) {
-        return new ParseState(ParseGraph.EMPTY, new ByteStreamSource(input), offset, new ImmutableList<>());
+        return new ParseState(ParseGraph.EMPTY, new ByteStreamSource(input), offset, new ImmutableList<>(), new ImmutableList<>());
     }
 
     public static ParseState createFromByteStream(final ByteStream input) {
         return createFromByteStream(input, ZERO);
     }
 
     public ParseState addBranch(final Token token) {
-        return new ParseState(order.addBranch(token), source, offset, token.isIterable() ? iterations.add(new ImmutablePair<>(token, ZERO)) : iterations);
+        return new ParseState(order.addBranch(token), source, offset, token.isIterable() ? iterations.add(new ImmutablePair<>(token, ZERO)) : iterations, references);
     }
 
     public ParseState closeBranch(final Token token) {
         if (token.isIterable() && !iterations.head.left.equals(token)) {
             throw new IllegalStateException(format("Cannot close branch for iterable token %s. Current iteration state is for token %s.", token.name, iterations.head.left.name));
         }
-        return new ParseState(order.closeBranch(), source, offset, token.isIterable() ? iterations.tail : iterations);
+        return new ParseState(order.closeBranch(), source, offset, token.isIterable() ? iterations.tail : iterations, references);
     }
 
     public ParseState add(final ParseValue parseValue) {
-        return new ParseState(order.add(parseValue), source, offset, iterations);
+        return new ParseState(order.add(parseValue), source, offset, iterations, references);
     }
 
     public ParseState add(final ParseReference parseReference) {
```

Comment:
Can't this the method where we actually add a parseReference to the references list? I know this is only used within Sub, but we could use this also within the CycleToken, I think.

Reply:
(This is a response to both your comments!) The problem is that not every invocation of `ParseState.add(...)` should lead to the creation of a `ParseReference` to add to the `references` collection. This is something specific to when a `CycleToken` is parsed, so that's why it happens there. For example a `Def` is not a `CycleToken` (because it cannot create a cycle) so adding a `Def` to the `ParseState` should not lead to adding something to the `references` collection.

I do totally agree though that the line in `CycleToken.parse(...)` looks quite ridiculous, so I will try to see if there is some way to shorten that in an intuitive manner that doesn't move `CycleToken`'s responsibility somewhere where it doesn't belong.

Reply:
I've mirrored the pattern we used on `Environment.addBranch(...)` and `ParseState.addBranch(...)`: there are now helper methods `Environment.addCycleReference(...)` and also `ParseState.addCycleReference(...)` that together compose what originally happened in `CycleToken.parse(...)`.

Reply:
We ended up improving this further by renaming the original `add(ParseReference)` to `createCycle(ParseReference)` and the `addCycleReference(...)` to just `add(...)` to better reflect what they actually do.